vignettes/targets_workflow.Rmd

```diff
@@ -0,0 +1,540 @@
+            description = glue::glue("Column classes: {col_class}")
+          ) %>%
+          add_results(report)
+      })
+      report
+    }
+
+    validation_violation <- function(report) {
+      violations_exist <- report$get_validations() %>%
+        summarise(
+          sum(num.violations, na.rm = TRUE) > 0
+        ) %>%
+        pull()
+      if (isTRUE(violations_exist)) {
+        abort(
+          "Validation schema error",
+          body = capture.output(report),
+          class = "validation_violation"
+        )
+      }
+      FALSE
+    }
+  })
+  file_move("_targets.R", "functions.R")
+})
+tar_script({
+  library(targets)
+  source("R/functions.R")
+  options(crayon.enabled = FALSE)
+  tar_option_set(
+    packages = c("readr", "dplyr", "data.validator", "assertr", "rlang",
+                 "purrr")
+  )
+  list(
+    tar_target(data_file, "data.csv", format = "file"),
+    tar_target(input_data, get_data(data_file)),
+    tar_target(schema_file, "schema.csv", format = "file"),
+    tar_target(schema, get_data(schema_file)),
+    tar_target(
+      metadata_validation,
+      metadata_validator(input_data, schema)
+    ),
+    tar_target(is_violation, validation_violation(metadata_validation))
+  )
+})
+```
+
+Analyzing metadata, such as the dimensions of a table or the presence of specific variables, is a crucial component of data validation. Typically, access to the complete dataset is needed in order to test metadata. `data.validator` has the `validate_if()` function that can confirm assertions on the data structure *itself*.
+
+Some typical examples are:
+
+1.  Check column names with `has_all_names()`, `has_only_names()`
+2.  Check the class of columns with `has_class()`
+3.  Check the dimensions of the table with a custom expression
+
+We need three functions for our simple metadata validation pipeline. The first, reads data and the schema from the CSV files that we wrote to disk in previous sections. The second validates the data structure by utilizing the stored schema file. The third checks the pipeline for violations and terminates in a controlled way by using a user-defined custom error class.
+
+The script that contains the functions for computing the pipeline "targets" reads:
+
+```{r, eval=FALSE}
+# R/functions.R
+get_data <- function(file) {
+  read_csv(file, col_types = cols())
+}
+
+metadata_validator <- function(data, schema) {
+  tar_assert_identical(names(schema), c("name", "type"))
+
+  report <- data_validation_report()
+  validate(data) %>%
+    validate_if(
+      has_all_names(!!!schema$name),
+      description = "Columns exist"
+    ) %>%
+    add_results(report)
+  walk(unique(schema$type), \(col_class) {
+    var_names <- schema %>%
+      filter(type == col_class) %>%
+      pull(name)
+    validate(data) %>%
+      validate_if(
+        has_class(!!!var_names, class = {{ col_class }}),
+        description = glue::glue("Column classes: {col_class}")
+      ) %>%
+      add_results(report)
```

Comment:
This is quite a lot for a lambda function, please consider extracting this part to a named function.

Reply:
I agree!